rfcs/2021-03-02-6600-vector-remap-language-web.md

```diff
@@ -0,0 +1,146 @@
+* The ability to share URLs for specific VRL events and/or programs. This would likely require using
+  a backend [web server] although it *may* be feasible to use URL hashes to achieve this.
+
+#### Prior art
+
+In PR [6590] I created a proof of concept for a VRL web app that uses the [Trunk] build tool and the
+[Yew] front-end framework for Rust, which is inspired by [React.js][react] and enables you to write
+web apps in pure Rust and without touching JavaScript. This POC does "work" but it creates an
+unacceptably large Wasm binary (~15 MB), largely due to it containing both VRL (which is bulky in
+itself) and *all* of the app's front-end logic. You can see a demo of that proof of concept
+[here][vrl_web_app].
+
+A much better solution, I believe, would be to create a web app that's mostly standard HTML, CSS,
+and JavaScript and delegating the VRL processing logic to one of the following:
+
+* A Wasm binary that only handles VRL logic
+* The [web server] proposed below
+
+### Web server
+
+A VRL web server would be a simple HTTP server with one endpoint to which you
+can `POST` the following:
+
+* A VRL program (as a string)
+* A VRL event (as JSON)
+* (optional) a compiler state. This would require one change to VRL:
+  `vrl::compiler::state::Compiler` would need to be serializable/deserializable
+  to/from JSON.
+
+Here's an example `POST` payload:
+
+```json
+{
+  "program": "del(.foo)\n.bar = baz",
+  "event": {
+    "foo": "bar"
+  },
+  "state": {
+    "variables": {
+      "baz": 45.2
+    }
+  }
+}
+```
+
+And the response:
+
+```json
+{
+  "event": {
+    "bar": 45.2
+  }
+}
+```
+
+A server of this kind would be quite germane to "serverless" environments like Lambda, as the logic
+is fully stateless.
+
+#### Implementation
+
+The [Warp] framework would be a natural candidate for this. Warp is the only widely used Rust web
+framework that both supports async and uses [Tokio] as its async runtime. The server would only need
+to have a single endpoint, perhaps `/compile`.
+
+#### Related possibilities
+
+* Run an instance of the server on a platform like Heroku
+* A `/functions` endpoint that describes the functions available in the current version
+* The ability to share URLs for specific events and/or programs. This would require either making
+  the server stateful and using some kind of persistent data store or devising a system of inferring
+  events/programs via hashes passed in via URL. It's likely that those URLs would grow quite
+  unwieldy, however, so we should experiment before taking this route.
+
+### Published module
+
+Yet another possibility worth considering would be publishing Wasm modules to hubs like [npm] and
+[Wasmer] that others can use in their own environments. Those modules could be released alongside
+versions of VRL. Tools from the Rust + Wasm ecosystem, like [wasm-pack], make this quite simple in
+principle. With published modules, users could run VRL
```

Comment:
Did the sentence end early or is a period missing?

Reply:
Excellent catch 👍🏼